tests/api_test.py

```diff
@@ -2863,6 +2863,38 @@ def clip_gradient(x):
 
     jax.grad(clip_gradient)(1.)  # doesn't crash
 
+  def test_nestable_vjp(self):
+    # Verify that https://github.com/google/jax/issues/3667 is resolved.
+    def f(x):
+        return x ** 2
+
+    @api.custom_vjp
+    def g(x):
+        return f(x)
+
+    def g_fwd(x):
+        y, f_vjp = api.vjp(f, x)
+        return y, f_vjp
+
+    def g_bwd(f_vjp, y_bar):
+        return f_vjp(y_bar)
+
+    g.defvjp(g_fwd, g_bwd)
+
+    # Check that VJP can be nested in simple situations.  For this to pass,
+    # vjp has to return a PyTree.
+    _, g_vjp = api.vjp(g, 1.0)
+    y, = g_vjp(1.0)
+    self.assertAllClose(y, jnp.array(2.0))
+
+    # Check that VJP can be nested in complex situations.  For this to pass,
+    # vjp can't treat the closed-over tracer x as a static argument.
```

Comment:
Is this static arg requirement a bug, for which we have an open issue? If so, it might be good to mention the issue url here in the comment, so we can track if/when it's fixed.

Reply:
I think this is what this pull request was supposed to fix?  In crafting this test case, I was looking for a simple example that put a tracer into the `consts` variable in `jax.interpreters.ad.vjp`.  That's where `x` ends up, so in the original code the tracer for `x` leaks when the `g_vjp` is returned from `z` because `jax.interpreters.ad.vjp` closes over the tracer.

I'm also happy to note that the issue in my own code where I nest VJPs is resolved by this change (but not by the simpler change originally proposed by shoyer), so it appears that VJPs are now nestable.